### DISCUSSION TITLE
Dependency Management

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#3454] [Q&A]
How do you stop the Oqtane License Component from being removed when you remove a module that uses it?

I have two module that use the License component. If I uninstall one of them then it removes the Lic Component and the system will not start 503, because there is a file missing.  That would be correct because the other component requires the Lic component.

**Comment by sbwalker** [comment]
This problem is not related to the License component - it is related to dependency management. If you have 2 modules which depend on assembly1, if you uninstall either of the modules the system will clean up ALL of the assets which were originally installed from the module's Nuget package - which means that it will remove assembly1. Oqtane does not currently contain any logic for tracking the dependency graph between various modules and the framework. 

Another scenario which is somewhat related is if you use the System Update capability to update the Oqtane framework, it will extract ALL assets from the Oqtane.Framework.Upgrade.zip. So if your modules are using versions of some assembles which are different than what is distributed with Oqtane, they will be overwritten by the System Update process.

**Reply by leigh-pointer** [reply]
@sbwalker ok I understand that. My root issue is the two separate external modules have been installed and both use the Oqtane License Component.  Uninstalling one the two removes, the Oqtane License Component.

**Reply by sbwalker** [reply]
@leigh-pointer in regards to the LicenseComponent, there are a couple of short-term options I can think of:

1. Hard-code some logic into the uninstall process which ignores the Oqtane.Licensing assemblies
2. Include Oqtane.Licensing by default in the Oqtane Framework distribution so that module developers do not need to include the package themselves.

I think the 2nd option makes more sense. In the longer term, Oqtane will need a way to track dependencies to handle these scenarios.

**Reply by leigh-pointer** [reply]
@sbwalker yes i agree, option 2 would be a lot better!

**Reply by sbwalker** [reply]
@leigh-pointer I was experimenting with adding the Oqtane.Licensing package to the framework and there is still a potential issue... if any module developer still bundles the DLLs with their module distribution and a user uninstalls their module it will remove the Oqtane.Licensing package - which affects any other modules which depend on it. The reason why a module developer may bundle the DLLs with their module are because they do not realize that the package is distributed with the framework, or perhaps they are targeting both Oqtane 4.x and 5.x with their module (and the package was not included with 4.x). So maybe option 1 is actually better  as a short-term solution - however Oqtane will definitely need a dependency manager soon (whcih is a non-trivial task based on all the scenarios which need to be handled).

**Reply by sbwalker** [reply]
Implemented option 1 in #3472 for 5.0 release

**Reply by leigh-pointer** [reply]
@sbwalker If we include the Oqtane.Licensing assemblies in the framework, and some logic into the **install** process which ignores the Oqtane.Licensing assemblies, as the framework would always have the most resent?  Or have you also explored that path?

**Reply by sbwalker** [reply]
@leigh-pointer you are correct - regardless of whether the Oqtane.Licensing assemblies are ignored during install or uninstall, they could still be distributed with the framework. This would make it easier for third party developers. However if you wanted your module or theme package to work with Oqtane 4.x and 5.x then you would still need to package the licensing assemblies so that they are installed on 4.x - and the uninstall would still be a problem. There is no backward compatible solution unfortunately - so I guess the best approach is to determine the optimal solution for 5.x.
